src/backend/distributed/utils/maintenanced.c

```diff
@@ -309,35 +296,54 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		 * tasks should do their own time math about whether to re-run checks.
 		 */
 
-		if (secondsSincePrevStatsCollection >= STATISTICS_COLLECTION_INTERVAL ||
-			(prevStatsCollectionFailed &&
-			 secondsSincePrevStatsCollection >= STATISTICS_COLLECTION_RETRY_INTERVAL))
-		{
 #ifdef HAVE_LIBCURL
-			if (EnableStatisticsCollection)
-			{
-				MemoryContext statsCollectionContext =
-					AllocSetContextCreate(CurrentMemoryContext,
-										  "StatsCollection",
-										  ALLOCSET_DEFAULT_MINSIZE,
-										  ALLOCSET_DEFAULT_INITSIZE,
-										  ALLOCSET_DEFAULT_MAXSIZE);
-				MemoryContext oldContext =
-					MemoryContextSwitchTo(statsCollectionContext);
+		if (EnableStatisticsCollection &&
+			GetCurrentTimestamp() >= nextStatsCollectionTime)
+		{
+			bool statsCollectionSuccess = false;
+			InvalidateMetadataSystemCache();
+			StartTransactionCommand();
 
+			/*
+			 * Lock the extension such that it cannot be dropped or created
+			 * concurrently. Skip statistics collection if citus extension is
+			 * not accessible.
+			 *
+			 * Similarly, we skip statistics collection if there exists any
+			 * version mismatch or the extension is not fully created yet.
+			 */
+			if (!LockCitusExtension())
```

Comment:
What do you think about differentiating between the failure to lock the extension and the failure to submit statistics? I feel like if the extension lock fails, retrying in a minute or so makes sense; if the stats submission fails, why retry in a minute? It seems like a good way to churn away hitting our stats server if it's having a brownout.

Reply:
Sounds good.